File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int solve(int idx, int buy, vector<int>& prices, vector<vector<int>>& dp){
        if(idx >= prices.size())
            return 0;
        
        if(dp[idx][buy] != -1){
            return dp[idx][buy];
        }
        
        if(buy == 1)
            return dp[idx][buy] = max(-prices[idx]+solve(idx+1, 0, prices, dp) , solve(idx+1, 1, prices, dp));
        
        
        return dp[idx][buy] = max(prices[idx] + solve(idx+2, 1, prices, dp), solve(idx+1, 0, prices, dp));
    }
    
    
    int maxProfit(vector<int>& prices) {
        vector<vector<int>> dp(prices.size(), vector<int>(2,-1));
        return solve(0,1,prices,dp);
    }
};

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    
    public int solve(int idx, int buy, int[] prices, int[][]dp){
        if(idx >= prices.length){
            return 0;
        }
        
        if(dp[idx][buy]!=-1){
            return dp[idx][buy];
        }
        
        int profit = 0;
        
        if(buy == 1){
            profit = Math.max(-prices[idx]+solve(idx+1,0,prices,dp) , 0 +solve(idx+1,1,prices,dp) );
        }else{
             profit = Math.max(prices[idx]+solve(idx+2,1,prices,dp) , 0 +solve(idx+1,0,prices,dp) );
        }
        return dp[idx][buy] = profit;
    }
    
    public int maxProfit(int[] prices) {
        int n = prices.length;
        
        int[][] dp = new int[n+1][2];
        
        for(int[] row : dp){
            Arrays.fill(row,-1);
        }
        
        return solve(0,1,prices,dp);
    }
}